File: src/main/java/forpleuvoir/hiirosakura/client/feature/regex/ChatMessageRegex.java
package forpleuvoir.hiirosakura.client.feature.regex;

import net.minecraft.text.Text;

import java.util.regex.Pattern;

/**
 * 聊天消息正则匹配
 *
 * @author forpleuvoir
 * <p>#project_name hiirosakura
 * <p>#package forpleuvoir.hiirosakura.client.feature.regex
 * <p>#class_name ChatMessageRegex
 * <p>#create_time 2021/6/23 23:36
 */
public class ChatMessageRegex {
    public static final String NAME = "name";
    public static final String MESSAGE = "message";
    private String name;
    private String message;

    public static ChatMessageRegex getInstance(Text text, String regex) {
        if (text != null && regex != null)
            return new ChatMessageRegex(text, regex);
        return null;
    }

    private ChatMessageRegex(Text text, String regex) {
        String content = preString(text);
        Pattern pattern = Pattern.compile(regex);
        var matcher = pattern.matcher(content);
        if (matcher.find()) {
            try {
                this.name = matcher.group(NAME);
                this.message = matcher.group(MESSAGE);
            } catch (Exception ignore) {
            }
        }
    }

    public String getPlayerName() {
        return this.name;
    }

    public String getMessage() {
        return this.message;
    }


    private String preString(Text text) {
        return text.getString().replaceAll("(§.)", "");
    }
}

File: src/main/kotlin/forpleuvoir/hiirosakura/client/feature/regex/ChatMessageRegex.kt
package forpleuvoir.hiirosakura.client.feature.regex

import forpleuvoir.hiirosakura.client.config.Configs
import net.minecraft.client.MinecraftClient
import net.minecraft.text.Text
import java.util.regex.Pattern

/**
 * 聊天消息正则匹配
 *
 *
 * 项目名 hiirosakura
 *
 * 包名 forpleuvoir.hiirosakura.client.feature.regex
 *
 * 文件名 ChatMessageRegex
 *
 * 创建时间 2021/6/23 23:36
 *
 *  @author forpleuvoir
 *
 */
class ChatMessageRegex private constructor(text: Text, regex: String) {
	var playerName: String? = null
	var message: String? = null

	private fun preString(text: Text): String {
		return text.string.replace(Regex("(§.)"), "")
	}

	companion object {
		private const val defaultRegex: String = "(<(?<name>(.*))>)\\s(?<message>.*)"
		private val currentServerAddress: String? get() = MinecraftClient.getInstance().currentServerEntry?.address

		@JvmStatic
		fun getInstance(text: Text): ChatMessageRegex {
			currentServerAddress?.let { address ->
				Configs.Values.CHAT_BUBBLE_REGEX[address]?.let {
					return ChatMessageRegex(text, it)
				}
			}
			return ChatMessageRegex(text, defaultRegex)
		}

		private const val NAME = "name"
		private const val MESSAGE = "message"
	}

	init {
		val content = preString(text)
		val pattern = Pattern.compile(regex)
		val matcher = pattern.matcher(content)
		if (matcher.find()) {
			try {
				playerName = matcher.group(NAME)
				message = matcher.group(MESSAGE)
			} catch (ignore: Exception) {
			}
		}
	}
}